### DISCUSSION TITLE
Download of files from S3 - performance

### DISCUSSION THREAD

**Discussion by DigNZ** [pocketbase/pocketbase#4332] [Q&A] [answered]
I have a server running pocketbase with large files stored in S3 and need to download a lot of them (its a video playlist) - does the data all flow through the pocketbase server or does pocketbase api/files endpoint hand it off somehow to be delivered from S3. Now I put this down in writing I'm not even sure if it makes sense.

We are noticing slower than expected downloads but there are a lot of factors and I'm not even sure that this is where optimisation needs to be focused but I would be interested to know how it currently works.

**Comment by ganigeorgiev** [comment, marked as answer]
All S3 operations are performed via the S3 APIs (including the object download), meaning that all requests are proxied through the PocketBase server.

> We are noticing slower than expected downloads...

Without details what do you mean by "slower than expected" and what is your basis for comparison I don't think I can help you. If you are comparing S3 CDN download vs S3 API download then the first one is expected to be faster.
Additionally S3 performance could vary for a lot of reasons - it depends on the S3 server resources, the location of the S3 and PocketBase servers, network speed of the host where PocketBase is deployed, network speed of the client when sending requests to PocketBase, whether the S3 server applies rate limiting or throttling, etc.
But in all cases S3 will be always slower than a the local filesystem (if that is what you are comparing with).

**Reply by DigNZ** [reply]
This actually answered my question perfectly fine. We use another platform that is S3 backed that uses signed urls for download directly from S3. But in this case it actually turns out the specific network situation had orders of magnitude larger impact on download times than pocketbase did. As long as we factor in that all the traffic does pass through the pocketbase server the performance impact is actually minimal.

**Comment by gokulgseven** [comment]
the data flows through the pocketbase server -> so basically your file is first downloaded on pocketbase server -> then downloaded by the user , 
in case of upload -> it is first uploaded to your pocketbase server and then uploaded to S3 

so basically you cant use the speed of s3 -> your bottleneck becomes the pocketbase server that you have deployed. 

i have been trying to figure out how to use presigned urls to increase speeds , havent had any luck so far

**Reply by ganigeorgiev** [reply]
I really don't know how to help and honestly I'm getting tired by this constant S3 related discussions as no one seems to bother elaborating how and what they actually measuring (what does _"speed of s3"_ even means?).

As mentioned earlier S3 upload/download performance could vary for a lot of reasons (_the distance between the servers probably would matter the most_) but without a concrete metric showing what actually you are measuring I really don't know how to help further.

There is no support for signed urls upload/download out of the box as it will complicate too much the regular record create/update flow and not all S3 providers support it.

The easiest way to "speed up" the S3 file downloads is to not use the S3 REST APIs and directly request the file from the S3 bucket public endpoint or CDN (eg. `https://yourcdn.com/:collectionId/:recordId/:filename`). The main drawbacks of that approach is that your bucket read access must be public (filenames are still random generated) and that thumbs will not be accessible if they don't exist because they are generated lazily on first `/api/files/*` thumb access.
